pandas/core/indexing.py

```diff
@@ -1727,6 +1728,44 @@ def _getitem_axis(self, key, axis: int):
         self._validate_key(key, axis)
         return self._get_label(key, axis=axis)
 
+    def _ensure_listlike_indexer(self, key, is_indexer_key: bool):
+        """
+        Ensure that a list-like of column labels are all present by adding them if
+        they do not already exist.
+
+        Parameters
+        ----------
+        key : _LocIndexer key or list-like of column labels
+            Target labels.
+        is_indexer_key : bool
+            Whether key is a _LocIndexer key
+        """
+        column_axis = 1
+        if is_indexer_key:
```

Comment:
do you *really* need is_indexer_key? this seems very odd, what exactly happens if you remove this keyword (specifically what about the below code is a problem)

Reply:
Well we need some way to differentiate whether `_ensure_listlike_indexer` is called from `DataFrame::_setitem_array` or `_LocIndexer::_get_setitem_indexer` because `key`s are in different formats depending on the caller. I have briefly explained this at https://github.com/pandas-dev/pandas/pull/29334#discussion_r359325599.

Notice that at if `is_indexer_key` is true, then I assign https://github.com/pandas-dev/pandas/blob/8ccd2b879b1a10beb7760f1f1b43fac36b41ae19/pandas/core/indexing.py#L1754 after some checks. So `is_indexer_key` is absolutely essential and the code below  the line doesn't make sense if `is_indexer_key` is removed.

Reply:
i think you could remove it if you *also* check is_list_like(key).

i just find this having to separate the keywords very hard to follow.

Reply:
`is_list_like` won't work as it returns `True` for tuple too. Checking `isinstance(key, tuple)` probably work but I'm not sure if it better reflects the intention.

Reply:
just exclude tuples

you *can* use self.name == 'loc' if you just want to handle that case.

Reply:
@jreback Currently an error is cause by the difference between `df` and `df.loc`. Eg.
```python
df = pd.DataFrame({"a": [1, 2], "b": [3, 4]}, index=["a", "b"])
>>> df["a"]  # column "a"
a    1
b    2
Name: a, dtype: int64
>>> df.loc["a"]  # row "a"
a    1
b    3
Name: a, dtype: int64
```
How to resolve this without explicitly indicating where `key` comes from?

Reply:
@jreback any suggestions?